Step2 - 지하철 노선 기능

tasklet1579 opened the pull request:
안녕하세요
2단계 지하철 노선 기능 제출합니다
인수 테스트를 먼저 작성하고 기능을 구현하라고 가이드 되어있는데 계속 번갈아가면서 하다보니 많이 헷갈렸던 것 같습니다 

잠자는 시간 빼고 하루 종일 컴퓨터 앞에 앉아 있는데 실제 집중하는 시간은 얼마 되지 않는 것 같습니다
이 과정에 참여한 것 자체만으로도 좋은 기회라고는 생각하지만 말씀하신데로 많은 것을 얻어가기 위해서는 조금 더 확실한 목표 설정이 필요한 것 같기도 하고.. 여러모로 잡 생각도 많아지는 시기 같습니다 
저는 시간을 내서 관악산에 한 번 다녀올까 생각합니다ㅎ
그럼 추가적인 개선 사항이 있을지 의견 부탁드리겠습니다
감사합니다!

tasklet1579 committed:
refactor : 지하철역 관련 공통 메서드 추출
- 생성
- 삭제
- 조회

tasklet1579 committed:
docs : 지하철 노선도 관리 문서 정리
- 지하철 노선 기능 추가

tasklet1579 committed:
refactor : 모든 테스트 케이스마다 시작하기 이전에 컨택스트를 재생성하고 포트 번호를 명시한다

tasklet1579 committed:
feat : 데이터베이스 테이블 초기화 컴포넌트 적용

tasklet1579 committed:
docs : 오타 수정

tasklet1579 committed:
test : application-test.properties

tasklet1579 committed:
feat : 지하철노선 기능 개발
- 지하철노선 생성
- 지하철노선 목록 조회
- 지하철노선 조회

tasklet1579 committed:
test : 지하철노선 기능 테스트
- 지하철노선 생성
- 지하철노선 목록 조회
- 지하철노선 조회

tasklet1579 committed:
test : 지하철노선 기능 인수 테스트
- 지하철노선 생성
- 지하철노선 목록 조회
- 지하철노선 조회

tasklet1579 committed:
feat : 지하철노선 리포지토리

tasklet1579 committed:
feat : 지하철노선 기능 개발
- 지하철노선 삭제

tasklet1579 committed:
test : 지하철노선 기능 인수 개발
- 지하철노선 삭제

tasklet1579 committed:
feat : 지하철노선 기능 인수 개발
- 지하철노선 수정

tasklet1579 committed:
test : 지하철노선 기능 인수 개발
- 지하철노선 수정

jaeyeonling commented on docs/지하철 노선도 관리.md:
```diff
@@ -88,3 +88,213 @@ Vary: Origin
+
+---
+
+### 2. 지하철 노선 기능
```
요구사항 작성 👍

jaeyeonling commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,65 @@
+        return LineResponse.of(line, toStations());
+    }
+
+    private static Long id = 1L;
+
+    private List<Station> toStations() {
+        return Arrays.asList(new Station(id, "지하철역_" + id++), new Station(id, "새로운지하철역_" + id++));
+    }
```
해당 코드는 어떠한 의미일까요..?
실제 Station을 사용하도록 수정해보면 좋겠어요.

jaeyeonling commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,57 @@
+    public LineResponse() {
+    }
+
+    public static LineResponse of(Line line, List<Station> stations) {
```
```
class A {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

클래스 구현 순서는 다음과 같습니다~
순서에 맞게 변경해주세요 :)

jaeyeonling commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public static LineResponse of(Line line, List<Station> stations) {
+        return new LineResponse(line.getId(), line.getName(), line.getColor(), new LinkedList<>(stations), line.getCreatedDate(), line.getModifiedDate());
```
120자가 넘는 것 같네요~
줄 바꿈을 통해 가독성을 높이면 좋겠습니다!

jaeyeonling commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @GetMapping(value = "/lines", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<LineResponse>> getLines() {
+        return ResponseEntity.ok().body(lineService.getLines());
```
```suggestion
        return ResponseEntity.ok(lineService.getLines());
```

소소하지만 다음과 같이 사용할 수 있습니다 ㅎㅎ

jaeyeonling commented on src/test/java/nextstep/subway/application/LineServiceTest.java:
```diff
@@ -0,0 +1,118 @@
+        LineResponse response = service.createLine(request);
+
+        // then
+        assertThat(response).isNotNull();
+        assertThat(response.getName()).isEqualTo("신분당선");
+        assertThat(response.getColor()).isEqualTo("bg-red-600");
```
단위테스트에는 하나의 assertion만 작성하는 것이 좋은데요.
테스트가 실패할 경우 이전 테스트 결과로 인한 전파가 발생했는지 단독적인 이슈인지를 직접 확인해봐야 하기 때문입니다.
그리고 테스트당 개념 하나만을 테스트해야 한다는 원칙도 위배되는 경우가 발생할 우려도 있습니다.
만약 여러 테스트가 필요한 경우라면 jUnit5에서 추가된 assertAll을 활용해주세요 :)

jaeyeonling commented on src/test/java/nextstep/subway/station/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,198 @@
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+
+        // then
+        assertThat(response.body().jsonPath().getString("id")).isNotNull();
+        assertThat(response.body().jsonPath().getList("name", String.class)).contains("신분당선");
+        assertThat(response.body().jsonPath().getString("stations")).isNotNull();
```
여러 값을 검증하는 경우 매번 jsonPath를 통해 만들지 않고 LineResponse로 만든 후 검증해도 괜찮겠네요!
`response.as(LineResponse.class)`를 활용해보세요~

jaeyeonling commented on src/test/java/nextstep/subway/station/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,198 @@
+        assertThat(response.body().jsonPath().getString("stations")).isNotNull();
+    }
+
+    ExtractableResponse<Response> getLine(Long id) {
```
해당 메서드는 라인 하나를 조회하는걸까요? 아니면 라인 전체를 조회하는걸까요?

jaeyeonling commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -141,35 +112,43 @@ void getStations() {
+    }
+
+    ExtractableResponse<Response> createStation(String name) {
+        Map<String, String> params = new HashMap<>();
+        params.put("name", name);
```
https://github.com/next-step/atdd-subway-admin/pull/613#discussion_r882214341 해당 코멘트를 확인해주세요.

jaeyeonling reviewed (changes requested):
미션을 잘 진행해주셨습니다 👍 
코멘트 남겨두었으니 확인 부탁드릴게요~
미션 이외에도 나누고 싶은 얘기가 있으시다면 편하게 DM 주세요 :)

tasklet1579 committed:
refactor : BaseEntity 클래스 접근 제어자 변경

tasklet1579 committed:
refactor : 2단계 리팩토링
- LineResponse : 클래스 내부 순서 변경
- Line : update 메서드 반환 값 삭제
- LineService : 메서드 네이밍 변경
- LineServiceTest : 테스트 코드 수정

tasklet1579 committed:
refactor : 2단계 리팩토링
- Station : 지하철노선 Id 필드 추가
- StationRepository : 지하철역 조회 메서드 추가
- LineService : 임시 지하철역 조회 로직 삭제
- LineController : 메서드 네이밍 수정

tasklet1579 committed:
test : 지하철노선 관련 기능 테스트

tasklet1579 committed:
refactor : 지하철역 조회 메서드 JPQL 적용

tasklet1579 commented:
안녕하세요~
저번에 주신 질문까지 모두 완료했습니다
저번에 주신 질문은 어떤 의도였는지 궁금하네요~
이번에도 잘 부탁드리겠습니다~!

tasklet1579 committed:
refactor : 테스트 패키지명 변경

jaeyeonling commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,62 @@
+    public LineResponse() {
+    }
+
+    public LineResponse(Long id, String name, String color, List<Station> stations, LocalDateTime createdDate, LocalDateTime modifiedDate) {
```
https://github.com/next-step/atdd-subway-admin/pull/639#discussion_r884134755 해당 코멘트 확인해주세요~

jaeyeonling commented on src/main/java/nextstep/subway/dto/LineRequest.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public Station getFirstStation(Long lineId) {
+        return new Station("지하철역_" + firstStationId, lineId);
```
`지하철역_`을 하드코딩 해두신 이유가 있을까요?
실제 서비스 관점에서 봤을 때 해당 시점에 station을 만들어주는게 맞을지도 같이 고민해보면 좋겠습니다!

jaeyeonling commented on src/test/java/nextstep/subway/acceptance/StationAcceptanceTest.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    ExtractableResponse<Response> createStation(String name) {
+        Map<String, String> params = new HashMap<>();
```
저번에 드린 질문의 의도를 물어보신 부분은 https://github.com/next-step/atdd-subway-admin/pull/639#discussion_r884135564 해당 내용이 맞으시겠죠?
실제 컨트롤러에서 받는 타입의 클래스를 사용하게 되면 타입안정성을 얻을 수 있다는 장점이 있다고 생각이 드는데요.
말씀하신 것과 같이 Map을 사용할 수 있을 것 같은데, Map을 사용하면 어떠한 장점이 있다고 생각하실까요?

jaeyeonling replied:
https://github.com/next-step/atdd-subway-admin/pull/639/files#r884496387 해당 코멘트는 놓치신걸까요?

jaeyeonling commented on src/main/java/nextstep/subway/domain/StationRepository.java:
```diff
@@ -1,10 +1,14 @@
 public interface StationRepository extends JpaRepository<Station, Long> {
     @Override
     List<Station> findAll();
+    @Query("SELECT s FROM Station s WHERE s.lineId = :lineId")
```
직접 쿼리를 작성해주신 이유가 있을까요?

jaeyeonling commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @DeleteMapping("/lines/{id}")
+    public ResponseEntity deleteLineById(@PathVariable Long id) {
```
제네릭에서 타입을 지정하지 않으면 어떠한 문제가 있을지 고민해보아요.

jaeyeonling replied:
https://github.com/next-step/atdd-subway-admin/pull/639/files#r884499956 해당 코멘트는 놓치신걸까요?

jaeyeonling commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,59 @@
+    public List<LineResponse> getLines() {
+        List<LineResponse> lines = new LinkedList<>();
+        for (Line line : lineRepository.findAll()) {
+            lines.add(LineResponse.of(line, stationRepository.findAllByLineId(line.getId())));
```
지금은 line의 갯수 만큼 해당 쿼리가 발생하는데요.
DB에 접근하는 IO는 비싼 자원으로, 접근을 최소화 해볼 수 있는 방법이 없을지 고민해보면 좋겠습니다.

jaeyeonling reviewed (changes requested):
미션을 잘 진행해주셨습니다~!
질문의 답변과 추가적인 코멘트 남겨두었으니 확인 부탁드려요!

tasklet1579 committed:
docs : 강의 자료

tasklet1579 committed:
style : 람다식 사용

tasklet1579 committed:
refactor : 생성자 수정

tasklet1579 committed:
refactor : 2단계 리팩토링
Station : 지하철노선 번호 매핑 로직 추가
StationRepository : @Query 삭제, 조회 메서드 추가
LineRequest : 지하철역 임시 로직 삭제
LineService : 지하철노선 생성 로직 수정
LineServiceTest : 초기화 데이터 추가

tasklet1579 committed:
test : 코드 정리

tasklet1579 committed:
refactor : Raw 타입 삭제

tasklet1579 commented:
안녕하세요~
2단계 미션 다시 확인 부탁드리겠습니다

jaeyeonling commented on src/main/java/nextstep/subway/domain/StationRepository.java:
```diff
@@ -1,10 +1,13 @@
     @Override
     List<Station> findAll();
+    List<Station> findAllByLineIdIsNotNull();
+    List<Station> findAllByLineId(@Param(value = "lineId")Long lineId);
```
```suggestion
    List<Station> findAllByLineId(Long lineId);
```

@Param은 생략해주셔도 될 것 같네요!

jaeyeonling commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,73 @@
+        List<LineResponse> lines = new LinkedList<>();
+        List<Station> stations = stationRepository.findAllByLineIdIsNotNull();
+        for (Line line : lineRepository.findAll()) {
+            lines.add(LineResponse.of(line, stations.stream()
```
매번 stream을 돌면서 station을 찾지 않고, Map의 자료구조를 사용해보면 좋겠어요~

jaeyeonling commented on src/test/java/nextstep/subway/application/LineServiceTest.java:
```diff
@@ -0,0 +1,113 @@
+    @Test
+    void createLine() {
+        // given
+        LineRequest request = new LineRequest("신분당선", "bg-red-600", 1L, 2L, 10);
```
1L, 2L은 어떠한 의미를 나타낼까요?

jaeyeonling reviewed (changes requested):
요구사항에 맞게 잘 반영해주셨네요!
코멘트 확인 부탁드리고 다음 단계 진행해주세요~
